src/selector-observer.js

```diff
@@ -0,0 +1,194 @@
+	 */
+	constructor (callback, options = {}) {
+		this.callback = callback;
+		this.options = options;
+
+		if (this.options.roots) {
+			for (let root of this.options.roots) {
+				this.observeRoot(root);
+			}
+		}
+		else {
+			this.observeRoot();
+		}
+	}
+
+	/**
+	 * Observe all selectors on a new root node.
+	 */
+	observeRoot (root = document) {
+		if (this.#roots.has(root)) {
+			return;
+		}
+
+		this.#roots.set(root, null);
+		this.#updateCSS(root);
+
+		root.addEventListener("transitionend", this, { capture: true });
+		root.addEventListener("animationend", this, { capture: true });
+	}
+
+	handleEvent (event) {
+		let { target } = event;
+		if (event.type === "transitionend") {
+			let selector = decode(event.propertyName.replace(/^--selector-observer-selector-/, ""));
+
+			console.log("transitionend", event.target, selector);
+			// this.callback(event.target);
+		}
+		else if (event.type === "animationend") {
+			let selector = decode(event.animationName.replace(/^selector-observer-/, ""));
+			console.log("animationend", event.target, selector);
+			// this.callback(event.target);
+		}
+	}
+	/**
+	 * Stop observing all selectors on a root node.
+	 */
+	unobserveRoot (root = document) {
+		if (!this.#roots.has(root)) {
+			return;
+		}
+
+		this.#roots.delete(root);
+
+		let sheet = this.#roots.get(root);
+
+		if (sheet) {
+			unadoptCSS(sheet, root);
+		}
+	}
+
+	/**
+	 * Observe a new selector on all roots.
+	 */
+	observe (selector, options) {
+		let immediate = options?.immediate ?? this.options.immediate;
+
+		if (immediate) {
+			let records = [];
+			for (let root of this.#roots.keys()) {
+				for (let target of root.querySelectorAll(selector)) {
+					let record = { root, target, selector, immediate: true };
+					records.push(record);
+				}
+			}
+
+			this.callback(records);
+		}
+
+		this.#selectors.add(selector);
+
+		// Add selector to CSS
+		for (let root of this.#roots.keys()) {
+			this.#updateCSS(root);
+		}
+	}
+
+	/**
+	 * Stop observing a selector on all roots.
+	 */
+	unobserve (selector) {
+		this.#selectors.delete(selector);
+
+		// Remove selector from CSS
+		for (let root of this.#roots.keys()) {
+			this.#updateCSS(root);
+		}
+	}
+
+	#updateCSS(root) {
+		if (this.#selectors.size === 0 && !this.#roots.get(root)) {
+			return;
+		}
+
+		let selectorList = Array.from(this.#selectors);
+		let ids = this.options.any ? ["any"] : selectorList.map(selector => encode(selector));
+
+		let animationNames = ids.map(id => "selector-observer-" + id);
+		let transitionProperties = ids.map(id => "--selector-observer-selector-" + id);
+
+		let css = `
+		${transitionProperties
+			.map(
+				property => `@property ${property} {
+			syntax: "<integer>";
+			initial-value: 0;
+			inherits: false;
+		}`,
+			)
+			.join("\n")}
+
+		${animationNames.map(animationName => `@keyframes ${animationName} {}`).join("\n")}
+
+		${selectorList.join(", ")} {
+			${transitionProperties.map(property => `${property}: 1`).join(";\n")};
```

Comment:
I don't fully understand why we set transition properties to `1` here if we do the same thing for every selector individually on line 159. 🤔

This appears to cause the `transitionend` event to fire for selectors that are not even matched (because they are included in a rule with selectors that have matched).

Reply:
By removing it, we seem to have the correct result in the test you wrote in `test.html`.

Reply:
Op, yes, this should definitely be removed

Reply:
```suggestion
```